src/code/z_map_exp.c

```diff
@@ -1,31 +1,255 @@
+        if (Play_GetOriginalSceneNumber(play->sceneNum) == sDungeonScenes[i]) {
+            return i;
+        }
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A430.s")
+    return -1;
+}
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A54C.s")
+/**
+ * Returns true if the current scene is a dungeon scene, false otherwise.
+ */
+s32 Map_IsInDungeonArea(PlayState* play) {
+    if (Map_GetDungeonAreaIndex(play) == -1) {
+        return false;
+    }
 
-#pragma GLOBAL_ASM("asm/non_matchings/code/z_map_exp/func_8010A580.s")
+    return true;
+}
+
+s32 sBossScenes[] = {
+    SCENE_MITURIN_BS, // DUNGEON_INDEX_WOODFALL_TEMPLE
+    SCENE_HAKUGIN_BS, // DUNGEON_INDEX_SNOWHEAD_TEMPLE
+    SCENE_SEA_BS,     // DUNGEON_INDEX_GREAT_BAY_TEMPLE
+    SCENE_INISIE_BS,  // DUNGEON_INDEX_STONE_TOWER_TEMPLE
+};
+
+/**
+ * If the current scene is a boss scene, this function returns an index
+ * indicating which one it is. Otherwise, it returns -1.
+ */
+s32 Map_GetBossAreaIndex(PlayState* play) {
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(sBossScenes); i++) {
+        if (Play_GetOriginalSceneNumber(play->sceneNum) == sBossScenes[i]) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+/**
+ * Returns true if the current scene is a boss scene, false otherwise.
+ */
+s32 Map_IsInBossArea(PlayState* play) {
+    if (Map_GetBossAreaIndex(play) == -1) {
+        return false;
+    }
+
+    return true;
+}
+
+s32 D_801BF5A4[] = {
+    SCENE_22DEKUCITY,
+    SCENE_KOEPONARACE,
+    SCENE_F01,
+};
+
+s32 func_8010A238(PlayState* play) {
+    s32 i;
+
+    for (i = 0; i < ARRAY_COUNT(D_801BF5A4); i++) {
+        if (Play_GetOriginalSceneNumber(play->sceneNum) == D_801BF5A4[i]) {
+            return i;
+        }
+    }
+
+    return -1;
+}
+
+s32 func_8010A2AC(PlayState* play) {
+    if (func_8010A238(play) == -1) {
+        return false;
+    }
+
+    return true;
+}
+
+void Map_SavePlayerInitialInfo(PlayState* play) {
+    Player* player = GET_PLAYER(play);
+
+    sPlayerInitialPosX = player->actor.world.pos.x;
+    sPlayerInitialPosZ = player->actor.world.pos.z;
+    sPlayerInitialDirection = (0x7FFF - player->actor.shape.rot.y) / 0x400;
+}
+
+void Map_InitRoomData(PlayState* play, s16 room) {
+    s32 mapIndex;
+    InterfaceContext* interfaceCtx;
+
+    mapIndex = gSaveContext.mapIndex;
+    func_80105C40(room);
+    if (room >= 0) {
+        if (Map_IsInDungeonOrBossArea(play)) {
+            interfaceCtx = &play->interfaceCtx;
+            gSaveContext.save.permanentSceneFlags[Play_GetOriginalSceneNumber(play->sceneNum)].unk_18 |=
+                gBitFlags[room];
+            interfaceCtx->mapRoomNum = room;
+            interfaceCtx->unk_27A = mapIndex;
+        }
+    } else {
+        play->interfaceCtx.mapRoomNum = 0;
+    }
+
+    if (gSaveContext.sunsSongState != SUNSSONG_SPEED_TIME) {
+        gSaveContext.sunsSongState = SUNSSONG_INACTIVE;
+    }
+}
```

Comment:
```suggestion
void Map_InitRoomData(PlayState* play, s16 room) {
    s32 mapIndex = gSaveContext.mapIndex;
    InterfaceContext* interfaceCtx = &play->interfaceCtx;

    func_80105C40(room);
    
    if (room >= 0) {
        if (Map_IsInDungeonOrBossArea(play)) {
            gSaveContext.save.permanentSceneFlags[Play_GetOriginalSceneNumber(play->sceneNum)].unk_18 |=
                gBitFlags[room];
            interfaceCtx->mapRoomNum = room;
            interfaceCtx->unk_27A = mapIndex;
        }
    } else {
        interfaceCtx->mapRoomNum = 0;
    }

    if (gSaveContext.sunsSongState != SUNSSONG_SPEED_TIME) {
        gSaveContext.sunsSongState = SUNSSONG_INACTIVE;
    }
}
```
This function should be structured like this (and it matches). `interfaceCtx` should be loaded at the top, and used where applicable